### ISSUE TITLE
Avoid including <windows.h>,  only HWND typedef is required

### ISSUE THREAD

**Issue by raysan5** [glfw/glfw#1348]
https://github.com/glfw/glfw/blob/cc805c09630a4b0b273cb5475a3bead706daa09d/include/GLFW/glfw3native.h#L92

`windows.h` defines lost of symbols that can result in conflicts with user defined symbols (i.e `Rectangle`, `CloseWindow`). This issue is easy to solve just defining the `HWND` type:
```c
typedef void *HWND;
```
For consistency with `windows.h` it can even be defined as:
```c
typedef void *PVOID;
typedef PVOID HANDLE;
typedef HANDLE HWND;
```
I can send a PR if you agree with this change integration.

**Comment by raysan5** [comment]
@tombsar I tried that change on my end for [raylib](https://github.com/raysan5/raylib) and everything seems to work fine.

**Comment by tombsar** [comment]
Ok, it looks like the only other thing needed in `glfw3native.h` itself is `HGLRC`, and that's conditional on `GLFW_EXPOSE_NATIVE_WGL` which you might not be using in raylib?.

I think we would have to continue to include `windows.h` (maybe with `WIN32_LEAN_AND_MEAN`?) in the case that `GLFW_EXPOSE_NATIVE_WGL` is enabled, but I like your suggestion otherwise.

**Comment by raysan5** [comment]
Oh! Yes! My mistake! You're right! I'm not using `GLFW_EXPOSE_NATIVE_WGL` so, don't have that problem.

`WIN32_LEAN_AND_MEAN` doesn't work for me, still exposes too many dup symbols.

I investigated a bit and `HGLRC`, it's most probably a `void *`, actually `glut.h` just defines `typedef void *HGLRC;`... but digging on `windows.h`, `windef.h` defines `DECLARE_HANDLE(HGLRC);` macro (also used for `DECLARE_HANDLE(HWND);`) that is defined in `winnt.h` as:
```c
typedef PVOID HANDLE;
#define DECLARE_HANDLE(name) typedef HANDLE name
```
Same file also defines `typedef void *PVOID;`.

In conclusion, most probably, those handlers are always `void *`.

**Comment by tombsar** [comment]
I think it would be cleanest to self-define HWND always, and pull in WGL stuff only if `GLFW_EXPOSE_NATIVE_WGL` is given. It might be possible to do that without including the entirety of windows.h, but I'm afraid I don't know (`wingdi.h`??). `WIN32_LEAN_AND_MEAN` would at least help to limit the fallout.

One other concern is how the header will interact with user code that wants to include `windows.h` for other reasons. If the user includes `glfw3native.h` before including `windows.h`, us having defined `HWND` already will probably cause errors, and visa versa. The main glfw header requires that you include it *after* `windows.h` already, but I don't think the documentation mentions this for glfw3native.h...

@elmindreda is this a change you would want to make? I don't use native access at all; maybe users are relying on it to include the whole of windows.h?

**Comment by raysan5** [comment]
About inclusion order of `glfw3native.h`, some security check could be added to avoid redefinition but, yes, in that case `windows.h` must be included before `glfw3native.h`.

Also, trying to get **Cocoa native window handler** (also defined as `void *`):
```c
    #if defined(__APPLE__)
        #include <unistd.h>             // Required for: usleep()
        #include <objc/message.h>       // Required for: objc_msgsend(), sel_registerName()
        
        #define GLFW_EXPOSE_NATIVE_COCOA
        #define GLFW_EXPOSE_NATIVE_NSGL
        #include <GLFW/glfw3native.h>   // Required for: glfwGetCocoaWindow(), glfwGetNSGLContext()
    #endif
```
I get the following compilation warning and error:
```
In file included from /Users/travis/build/raysan5/raylib/src/external/glfw/include/GLFW/glfw3.h:195:
/System/Library/Frameworks/OpenGL.framework/Headers/gl.h:5:2: warning: gl.h and gl3.h are both included. Compiler will not invoke errors if using removed OpenGL functionality. [-W#warnings]
#warning gl.h and gl3.h are both included.  Compiler will not invoke errors if using removed OpenGL functionality.
 ^
In file included from /Users/travis/build/raysan5/raylib/src/core.c:171:
/Users/travis/build/raysan5/raylib/src/external/glfw/include/GLFW/glfw3native.h:104:17: error: typedef redefinition with different types ('void *' vs 'struct objc_object *')
  typedef void* id;
                ^
/usr/include/objc/objc.h:46:29: note: previous definition is here
typedef struct objc_object *id;
```

**Comment by tombsar** [comment]
The warning implies that you have something in your code that includes "gl3.h" (maybe an OpenGL loader library?). Defining `GLFW_INCLUDE_NONE` before including glfw.h is the intended way to disable that second inclusion, which should fix the warning.

The error arises because you include `objc/message.h`, which in turn includes `objc.h`, and defines `id` properly. I believe the error could be avoided if we guard glfw3native.h's dummy typedef with `#if !OBJC_TYPES_DEFINED`. That's really a separate issue/PR (as it affects MacOS only), although the principle is the same as we were discussing for Windows.

**Comment by raysan5** [comment]
Thanks! Effectively `gl.h` was already included by a secondary library (`glad.h`).

Also tried the guard around the typedef for MacOS but it keeps breaking... should I open a second related issue?

**Comment by SamuelMarks** [comment]
The `Windows.h` header defined `_WINDOWS_` so you can always guard against redefinitions if others are including that header by doing an `ifdef` against that symbol.
